### PR TITLE
Support nulls in List type

### LINKED ISSUE #56: Reading COCO segmentation failed due to nulls
# Problem Statement

Reading COCO dataset with segmentation, failed with error message:

```
row_id=12472 Mismatching child array lengths
```

Addining more debug info into `FileReader::GetListArray()` and `FileReader::GetStructArray()`, I got the following messages:

```
Failed to build struct array: field=segmentation children=["height", "polygon", "rle", "type", "width"] batch=17: Mismatching child array lengths
Arrays: [[
  478,
  478,
  478,
  478,
  478,
  478,
  478,
  478,
  478,
  478,
  ...
  478,
  478,
  478,
  478,
  478,
  478,
  478,
  478,
  478,
  478
], <Invalid array: Length spanned by list offsets (23) larger than values array (length 22)>, [
  [],
...
```

### PR DESCRIPTION
**Pull request by eddyxu**
* Set nulls during the read list process. 
* Closes #56, COCO query failed due to optional Segmentation Polygon / RLE structure.

### CODE REVIEW

File: cpp/src/lance/io/reader.cc [hunk 1]
```diff
@@ -156,19 +151,14 @@ ::arrow::Result<::std::shared_ptr<::arrow::Scalar>> FileReader::GetListScalar(
   decoder->Reset(pos, length);
   ARROW_ASSIGN_OR_RAISE(auto offsets_arr, decoder->ToArray(idx, 2));
   auto offsets = std::static_pointer_cast<::arrow::Int32Array>(offsets_arr);
+  if (offsets->Value(0) == offsets->Value(1)) {
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.cc:154]: so we're not using the null bitmap to check?

File: cpp/src/lance/io/reader.cc [hunk 2]
```diff
@@ -347,26 +336,33 @@ ::arrow::Result<std::shared_ptr<::arrow::Array>> FileReader::GetListArray(
   auto length = params.length;
   auto start = params.offset.value();
 
-  ARROW_ASSIGN_OR_RAISE(auto offsets_arr, GetPrimitiveArray(field, batch_id, params));
+  std::shared_ptr<::arrow::Array> offsets_arr;
+  if (length.has_value()) {
+    ARROW_ASSIGN_OR_RAISE(
+        offsets_arr,
+        GetPrimitiveArray(field, batch_id, ArrayReadParams(start, length.value() + 1)));
```
Comment (changhiskhan) [inline cpp/src/lance/io/reader.cc:343]: why +1 here?

**Review by changhiskhan — COMMENTED** [review]
minor questions

File: cpp/src/lance/io/reader.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.cc:154]: Null bitmap is not actually stored physically. It only exists during in-memory reconstruction.

File: cpp/src/lance/io/reader.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/io/reader.cc:343]: So for a N-element array, the offsets has N+1 values, i.e. `0, offset0, offset1, .... offsetN`.

**Review by changhiskhan — APPROVED** [review]
thx for clarifying

**Pull request merged**